File: Data Structures & Algorithms/max-water-container/submission-1.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int l = 0, r = heights.size() - 1, max = 0;
        while (l < r) {
            if (heights[l] < heights[r]) {
                max = std::max({max, heights[l] * (r - l)});
                ++l;
            }
            else {
                max = std::max({max, heights[r] * (r - l)});
                --r;
            }
        }
        return max;
    }
};

File: Data Structures & Algorithms/eating-bananas/submission-1.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        def cal_time(banana, speed):
            return (banana - 1) // speed + 1

        max_banana = max(piles)
        left, right = 1, max_banana + 1
        while left < right:
            mid = (left + right) // 2
            time_needed = sum(cal_time(x, mid) for x in piles)
            if time_needed <= h:
                right = mid
            else:
                left = mid + 1
        return left